datasource/src/main/scala/quasar/physical/mongo/decoder.scala

```diff
@@ -75,30 +69,19 @@ object decoder {
       case BsonType.INT64 => QLong
       case BsonType.DOUBLE => QDouble
       case BsonType.DECIMAL128 => {
-
         val dec128: Decimal128 = bson.asDecimal128().getValue()
-        if (dec128.isNaN()) {
-          QNull
-        } else if (dec128.isInfinite()) {
-          QNull
-        } else try {
-          val decimal: JDecimal = dec128.bigDecimalValue()
-          if (decimal.scale() < 1 && decimal.compareTo(maxLong) < 1 && decimal.compareTo(minLong) > -1) {
-            QLong
-          } else {
-            if (decimal.abs().compareTo(tinyDbl) < 0) {
-              QReal
-            } else
-              if (decimal.compareTo(maxDbl) < 1 && decimal.compareTo(minDbl) > -1) {
-                QDouble
-              } else {
-                QReal
-              }
-          }
+        if (dec128.isNaN()) QNull
+        else if (dec128.isInfinite()) QNull
+        else try {
+          val decimal: BigDecimal = BigDecimal(dec128.bigDecimalValue())
+          if (decimal.isValidLong) QLong
+          else if (decimal.isDecimalDouble) QDouble
+          else QReal
         } catch {
           case e: ArithmeticException => QLong
```

Comment:
This is for `-0`, right?  I'd prefer this to be coded in an explicit `if/then/else` branch rather than a `try/catch` block

Reply:
Me too, but there is no api for it. I.e. you can't know that this is -0 without trying to convert. Equality with constant won't work too because decimal128 has tons of negative zeros